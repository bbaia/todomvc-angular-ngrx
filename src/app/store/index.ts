export * from './meta-reducers';
export * from './reducers';
export * from './router-selectors';
export * from './router-state-serializer';
