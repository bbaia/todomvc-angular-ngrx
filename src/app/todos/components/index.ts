export * from './footer/footer.component';
export * from './new-todo/new-todo.component';
export * from './todo-list-item/todo-list-item.component';
export * from './todo-list/todo-list.component';
