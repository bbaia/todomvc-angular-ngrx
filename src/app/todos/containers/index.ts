export * from './todo/todo.component';
