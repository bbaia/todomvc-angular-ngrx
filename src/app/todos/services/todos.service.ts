import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { Todo } from '../models';

@Injectable()
export class TodosService {
  constructor(private http: HttpClient) {}

  getTodos(): Observable<Todo[]> {
    return this.http.get<any[]>('assets/todos.json').pipe(
      map(data =>
        data.map<Todo>(x => ({
          id: x.id,
          text: x.text,
          creationDate: new Date(x.creationDate),
          completed: x.completed,
        })),
      ),
    );
  }
}
