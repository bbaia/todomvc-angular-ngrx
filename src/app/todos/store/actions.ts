import { createAction, props } from '@ngrx/store';
import { Todo, TodoFilter } from '../models';

export const addAction = createAction('[TODO] add', (text: string) => ({
  id: Math.random(),
  text,
}));

export const loadAction = createAction('[TODO] load');

export const loadSuccessAction = createAction(
  '[TODO] load success',
  props<{ todos: Todo[] }>(),
);

export const deleteAction = createAction(
  '[TODO] delete',
  props<{ id: number }>(),
);

export const toggleAction = createAction(
  '[TODO] toggle',
  props<{ id: number }>(),
);

export const updateAction = createAction(
  '[TODO] update',
  props<{ id: number; text: string }>(),
);

export const clearCompletedAction = createAction('[TODO] clear completed');

export const setFilterAction = createAction(
  '[TODO] Set filter',
  props<{ filter: TodoFilter }>(),
);
