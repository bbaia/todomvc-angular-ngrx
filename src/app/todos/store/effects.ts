import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { map, switchMap, tap } from 'rxjs/operators';
import { TodosService } from '../services';
import * as fromTodos from '../store';

@Injectable()
export class TodosEffects {
  loadTodos$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(fromTodos.loadAction),
      switchMap(action => this.todosService.getTodos()),
      map(todos => fromTodos.loadSuccessAction({ todos })),
      // delay(1000), // Simulate network latency for loading animation
    );
  });

  filter$ = createEffect(
    () => {
      return this.actions$.pipe(
        ofType(fromTodos.setFilterAction),
        tap(action => {
          switch (action.filter) {
            case 'SHOW_ACTIVE': {
              this.router.navigate(['/', 'active']);
              break;
            }
            case 'SHOW_COMPLETED': {
              this.router.navigate(['/', 'completed']);
              break;
            }
            default: {
              this.router.navigate(['/', 'all']);
              break;
            }
          }
        }),
      );
    },
    { dispatch: false },
  );

  constructor(
    private actions$: Actions,
    private router: Router,
    private todosService: TodosService,
  ) {}
}
