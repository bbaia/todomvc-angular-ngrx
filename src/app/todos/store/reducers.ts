import { createReducer, on } from '@ngrx/store';
import * as fromTodos from './actions';
import * as todoEntity from './entities/todo';

export interface State {
  data: todoEntity.State;
  loading: boolean;
}

export const initialState: State = {
  data: todoEntity.initialState,
  loading: false,
};

export const reducer = createReducer(
  initialState,
  on(
    fromTodos.addAction,
    (state, action): State => ({
      ...state,
      data: todoEntity.adapter.addOne(
        {
          id: action.id,
          text: action.text,
          creationDate: new Date(),
          completed: false,
        },
        state.data,
      ),
    }),
  ),
  on(
    fromTodos.loadAction,
    (state, action): State => ({
      ...state,
      loading: true,
    }),
  ),
  on(
    fromTodos.loadSuccessAction,
    (state, action): State => ({
      ...state,
      data: todoEntity.adapter.setAll(action.todos, state.data),
      loading: false,
    }),
  ),
  on(
    fromTodos.toggleAction,
    (state, action): State => ({
      ...state,
      data: todoEntity.adapter.updateOne(
        {
          id: action.id,
          changes: {
            completed: !state.data.entities[action.id]?.completed,
          },
        },
        state.data,
      ),
    }),
  ),
  on(
    fromTodos.deleteAction,
    (state, action): State => ({
      ...state,
      data: todoEntity.adapter.removeOne(action.id, state.data),
    }),
  ),
  on(
    fromTodos.updateAction,
    (state, action): State => ({
      ...state,
      data: todoEntity.adapter.updateOne(
        { id: action.id, changes: { text: action.text } },
        state.data,
      ),
    }),
  ),
  on(
    fromTodos.clearCompletedAction,
    (state, action): State => ({
      ...state,
      data: todoEntity.adapter.removeMany(
        state.data.ids.filter(id => state.data.entities[id]?.completed),
        state.data,
      ),
    }),
  ),
);
