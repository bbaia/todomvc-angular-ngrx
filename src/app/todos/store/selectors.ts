import { createFeatureSelector, createSelector } from '@ngrx/store';
import * as fromApp from '../../store';
import { TodoFilter } from '../models';
import * as todoEntity from './entities/todo';
import * as fromTodos from './reducers';

export const selectTodoState = createFeatureSelector<fromTodos.State>('todos');

// Raw selectors

const { selectIds, selectEntities, selectAll, selectTotal } =
  todoEntity.adapter.getSelectors();

export const selectAllTodos = createSelector(selectTodoState, state =>
  selectAll(state.data),
);

export const selectTotalTodos = createSelector(selectTodoState, state =>
  selectTotal(state.data),
);

export const selectLoading = createSelector(
  selectTodoState,
  state => state.loading,
);

// Calculated selectors

export const selectFilter = createSelector(
  fromApp.selectRouteParamFilter,
  (routeFilter): TodoFilter => {
    switch (routeFilter) {
      case 'active': {
        return 'SHOW_ACTIVE';
      }
      case 'completed': {
        return 'SHOW_COMPLETED';
      }
      default: {
        return 'SHOW_ALL';
      }
    }
  },
);

export const selectFilteredTodos = createSelector(
  selectAllTodos,
  selectFilter,
  (todos, filter) => {
    switch (filter) {
      default:
      case 'SHOW_ALL':
        return todos;
      case 'SHOW_COMPLETED':
        return todos.filter(t => t.completed);
      case 'SHOW_ACTIVE':
        return todos.filter(t => !t.completed);
    }
  },
);

export const selectHasTodos = createSelector(selectTotalTodos, totalTodos => {
  return totalTodos > 0;
});

export const selectHasCompletedTodos = createSelector(selectAllTodos, todos => {
  return todos.filter(t => t.completed).length > 0;
});

export const selectUndoneTodosCount = createSelector(selectAllTodos, todos => {
  return todos.filter(t => !t.completed).length;
});
