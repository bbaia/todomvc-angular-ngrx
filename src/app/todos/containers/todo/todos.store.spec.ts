import { TestBed } from '@angular/core/testing';
import { skip, take, tap } from 'rxjs/operators';
import { TestScheduler } from 'rxjs/testing';
import { Todo } from '../../models';
import { TodosService } from '../../services';
import { TodosStore } from './todos.store';

describe('TodosStore', () => {
  let todosService: TodosService;
  let todosStore: TodosStore;

  const testTodos: Todo[] = [
    {
      id: 0.1,
      text: 'Learn Angular',
      creationDate: new Date('2020-01-01'),
      completed: true,
    },
    {
      id: 0.2,
      text: 'Learn NgRx',
      creationDate: new Date('2020-02-01'),
      completed: true,
    },
    {
      id: 0.3,
      text: 'Learn Jest',
      creationDate: new Date('2020-03-01'),
      completed: false,
    },
  ];

  beforeEach(() => {
    TestBed.configureTestingModule({
      providers: [
        TodosStore,
        {
          provide: TodosService,
          useValue: { getTodos: jest.fn() },
        },
      ],
    });

    todosService = TestBed.inject(TodosService);
    todosStore = TestBed.inject(TodosStore);
  });

  describe('Effects', () => {
    let testScheduler: TestScheduler;

    beforeEach(() => {
      testScheduler = new TestScheduler((actual, expected) => {
        expect(actual).toEqual(expected);
      });
    });

    it('should load todos', done => {
      testScheduler.run(
        ({ cold, hot, expectObservable, expectSubscriptions, flush }) => {
          // Arrange
          todosService.getTodos = jest.fn(() => cold('--a|', { a: testTodos }));

          // Act
          todosStore.load();

          // Assert
          expectObservable(todosStore.totalTodos$).toBe('a-b', { a: 0, b: 3 });
          todosStore.state$
            .pipe(
              take(1),
              tap(state => expect(state).toMatchSnapshot()),
            )
            .subscribe();
          todosStore.state$
            .pipe(
              skip(1),
              tap(state => {
                expect(state).toMatchSnapshot();
                done();
              }),
            )
            .subscribe();
        },
      );
    });
  });

  describe('Updaters', () => {
    beforeEach(() => {
      jest.useFakeTimers();
      jest.setSystemTime(new Date('2020-01-01'));
    });

    afterEach(() => {
      jest.useRealTimers();
    });

    it('should add todos', done => {
      // Arrange
      jest.spyOn(Math, 'random').mockReturnValue(0.1);

      // Act
      todosStore.add('Learn Angular');
      jest.spyOn(Math, 'random').mockRestore();

      // Assert
      todosStore.state$.subscribe(state => {
        expect(state.todos.length).toEqual(1);
        expect(state).toMatchSnapshot();
        done();
      });
    });

    it('should mark todos as complete', done => {
      // Arrange
      todosStore.patchState({
        todos: [
          {
            id: 0.07,
            text: 'Learn Jest',
            creationDate: new Date(),
            completed: false,
          },
        ],
      });

      // Act
      todosStore.toggle(0.07);

      // Assert
      todosStore.state$.subscribe(state => {
        expect(state).toMatchSnapshot();
        done();
      });
    });

    it('should delete todos', done => {
      // Arrange
      todosStore.patchState({
        todos: [
          {
            id: 0.07,
            text: 'Learn Jest',
            creationDate: new Date(),
            completed: false,
          },
        ],
      });

      // Act
      todosStore.delete(0.07);

      // Assert
      todosStore.state$.subscribe(state => {
        expect(state).toMatchSnapshot();
        done();
      });
    });

    it('should update todos', done => {
      // Arrange
      todosStore.patchState({
        todos: [
          {
            id: 0.07,
            text: 'Learn Jest',
            creationDate: new Date(),
            completed: false,
          },
        ],
      });

      // Act
      todosStore.update({
        id: 0.07,
        text: 'Learn Angular',
      });

      // Assert
      todosStore.state$.subscribe(state => {
        expect(state).toMatchSnapshot();
        done();
      });
    });

    it('should clear completed todos', done => {
      // Arrange
      todosStore.patchState({
        todos: [
          {
            id: 0.01,
            text: 'Learn Angular',
            creationDate: new Date(),
            completed: true,
          },
          {
            id: 0.02,
            text: 'Learn NgRx',
            creationDate: new Date(),
            completed: true,
          },
          {
            id: 0.03,
            text: 'Learn Jest',
            creationDate: new Date(),
            completed: false,
          },
        ],
      });

      // Act
      todosStore.clearCompleted();

      // Assert
      todosStore.state$.subscribe(state => {
        expect(state).toMatchSnapshot();
        done();
      });
    });
  });

  describe('Selectors', () => {
    it('filteredTodos$ should return all todos', done => {
      todosStore.patchState({
        todos: testTodos,
        filter: 'SHOW_ALL',
      });
      todosStore.filteredTodos$.subscribe(todos => {
        expect(todos.length).toEqual(3);
        expect(todos).toMatchSnapshot('All todos');
        done();
      });
    });

    it('filteredTodos$ should return completed todos', done => {
      todosStore.patchState({
        todos: testTodos,
        filter: 'SHOW_COMPLETED',
      });
      todosStore.filteredTodos$.subscribe(todos => {
        expect(todos.length).toEqual(2);
        expect(todos).toMatchSnapshot('Completed todos');
        done();
      });
    });

    it('filteredTodos$ should return active todos', done => {
      todosStore.patchState({
        todos: testTodos,
        filter: 'SHOW_ACTIVE',
      });
      todosStore.filteredTodos$.subscribe(todos => {
        expect(todos.length).toEqual(1);
        expect(todos).toMatchSnapshot('Active todos');
        done();
      });
    });

    it('hasTodos$ should return true with todos', done => {
      todosStore.patchState({
        todos: testTodos,
        filter: 'SHOW_ACTIVE',
      });
      todosStore.hasTodos$.subscribe(result => {
        expect(result).toEqual(true);
        done();
      });
    });

    it('hasTodos$ should return false with no todos', done => {
      todosStore.hasTodos$.subscribe(result => {
        expect(result).toEqual(false);
        done();
      });
    });

    it('hasCompletedTodos$ should return true with completed todos', done => {
      todosStore.patchState({
        todos: testTodos,
      });
      todosStore.hasCompletedTodos$.subscribe(result => {
        expect(result).toEqual(true);
        done();
      });
    });

    it('hasCompletedTodos$ should return false with only undone todos', done => {
      todosStore.patchState({
        todos: [testTodos[2]],
      });
      todosStore.hasCompletedTodos$.subscribe(result => {
        expect(result).toEqual(false);
        done();
      });
    });

    it('undoneTodosCount$ should return undone todos count', done => {
      todosStore.patchState({
        todos: testTodos,
      });
      todosStore.undoneTodosCount$.subscribe(count => {
        expect(count).toEqual(1);
        done();
      });
    });
  });
});
