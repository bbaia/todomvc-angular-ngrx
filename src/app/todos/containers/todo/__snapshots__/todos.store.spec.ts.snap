// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodosStore Effects should load todos 1`] = `
{
  "filter": "SHOW_ALL",
  "loading": true,
  "todos": [],
}
`;

exports[`TodosStore Effects should load todos 2`] = `
{
  "filter": "SHOW_ALL",
  "loading": false,
  "todos": [
    {
      "completed": true,
      "creationDate": 2020-01-01T00:00:00.000Z,
      "id": 0.1,
      "text": "Learn Angular",
    },
    {
      "completed": true,
      "creationDate": 2020-02-01T00:00:00.000Z,
      "id": 0.2,
      "text": "Learn NgRx",
    },
    {
      "completed": false,
      "creationDate": 2020-03-01T00:00:00.000Z,
      "id": 0.3,
      "text": "Learn Jest",
    },
  ],
}
`;

exports[`TodosStore Selectors filteredTodos$ should return active todos: Active todos 1`] = `
[
  {
    "completed": false,
    "creationDate": 2020-03-01T00:00:00.000Z,
    "id": 0.3,
    "text": "Learn Jest",
  },
]
`;

exports[`TodosStore Selectors filteredTodos$ should return all todos: All todos 1`] = `
[
  {
    "completed": true,
    "creationDate": 2020-01-01T00:00:00.000Z,
    "id": 0.1,
    "text": "Learn Angular",
  },
  {
    "completed": true,
    "creationDate": 2020-02-01T00:00:00.000Z,
    "id": 0.2,
    "text": "Learn NgRx",
  },
  {
    "completed": false,
    "creationDate": 2020-03-01T00:00:00.000Z,
    "id": 0.3,
    "text": "Learn Jest",
  },
]
`;

exports[`TodosStore Selectors filteredTodos$ should return completed todos: Completed todos 1`] = `
[
  {
    "completed": true,
    "creationDate": 2020-01-01T00:00:00.000Z,
    "id": 0.1,
    "text": "Learn Angular",
  },
  {
    "completed": true,
    "creationDate": 2020-02-01T00:00:00.000Z,
    "id": 0.2,
    "text": "Learn NgRx",
  },
]
`;

exports[`TodosStore Updaters should add todos 1`] = `
{
  "filter": "SHOW_ALL",
  "loading": false,
  "todos": [
    {
      "completed": false,
      "creationDate": 2020-01-01T00:00:00.000Z,
      "id": 0.1,
      "text": "Learn Angular",
    },
  ],
}
`;

exports[`TodosStore Updaters should clear completed todos 1`] = `
{
  "filter": "SHOW_ALL",
  "loading": false,
  "todos": [
    {
      "completed": false,
      "creationDate": 2020-01-01T00:00:00.000Z,
      "id": 0.03,
      "text": "Learn Jest",
    },
  ],
}
`;

exports[`TodosStore Updaters should delete todos 1`] = `
{
  "filter": "SHOW_ALL",
  "loading": false,
  "todos": [],
}
`;

exports[`TodosStore Updaters should mark todos as complete 1`] = `
{
  "filter": "SHOW_ALL",
  "loading": false,
  "todos": [
    {
      "completed": true,
      "creationDate": 2020-01-01T00:00:00.000Z,
      "id": 0.07,
      "text": "Learn Jest",
    },
  ],
}
`;

exports[`TodosStore Updaters should update todos 1`] = `
{
  "filter": "SHOW_ALL",
  "loading": false,
  "todos": [
    {
      "completed": false,
      "creationDate": 2020-01-01T00:00:00.000Z,
      "id": 0.07,
      "text": "Learn Angular",
    },
  ],
}
`;
